File: devicesetup/src/main/java/io/particle/android/sdk/devicesetup/setupsteps/ConfigureAPStep.java
package io.particle.android.sdk.devicesetup.setupsteps;


import android.content.Context;

import java.io.IOException;
import java.security.PublicKey;

import io.particle.android.sdk.devicesetup.commands.CommandClient;
import io.particle.android.sdk.devicesetup.commands.ConfigureApCommand;
import io.particle.android.sdk.devicesetup.commands.InterfaceBindingSocketFactory;
import io.particle.android.sdk.devicesetup.commands.ScanApCommand;
import io.particle.android.sdk.devicesetup.commands.data.WifiSecurity;
import io.particle.android.sdk.utils.Crypto;


public class ConfigureAPStep extends SetupStep {

    private final CommandClient commandClient;
    private final ScanApCommand.Scan networkToConnectTo;
    private final String networkSecretPlaintext;
    private final PublicKey publicKey;
    private final Context ctx;

    private volatile boolean commandSent = false;

    public ConfigureAPStep(StepConfig stepConfig, CommandClient commandClient,
                           ScanApCommand.Scan networkToConnectTo, String networkSecretPlaintext,
                           PublicKey publicKey, Context ctx) {
        super(stepConfig);
        this.commandClient = commandClient;
        this.networkToConnectTo = networkToConnectTo;
        this.networkSecretPlaintext = networkSecretPlaintext;
        this.publicKey = publicKey;
        this.ctx = ctx;
    }

    protected void onRunStep() throws SetupStepException {
        WifiSecurity wifiSecurity = WifiSecurity.fromInteger(networkToConnectTo.wifiSecurityType);
        ConfigureApCommand.Builder builder = ConfigureApCommand.newBuilder()
                .setSsid(networkToConnectTo.ssid)
                .setSecurityType(wifiSecurity)
                .setChannel(networkToConnectTo.channel)
                .setIdx(0);
        if (wifiSecurity != WifiSecurity.OPEN) {
            try {
                builder.setEncryptedPasswordHex(
                        Crypto.encryptAndEncodeToHex(networkSecretPlaintext, publicKey));
            } catch (Crypto.CryptoException e) {
                // FIXME: try to throw a more specific exception here.
                // Don't throw SetupException here -- if this is failing, it's not
                // going to get any better by the running this SetupStep again, and
                // it can really only fail if the surrounding app code is doing something
                // wrong.  To wit: you *want* the app to crash here (or at least
                // throw out a dialog saying "horrible thing happened!  horrible error
                // code: ..." and then return to a safe "default" activity.
                throw new RuntimeException("Error encrypting network credentials", e);
            }
        }
        ConfigureApCommand command = builder.build();

        try {
            ConfigureApCommand.Response response = commandClient.sendCommandAndReturnResponse(
                    command, ConfigureApCommand.Response.class, new InterfaceBindingSocketFactory(ctx));
            if (!response.isOk()) {
                throw new SetupStepException("Error response code " + response.responseCode +
                        " while configuring device");
            }
            log.d("Configure AP command returned: " + response.responseCode);
            commandSent = true;

        } catch (IOException e) {
            throw new SetupStepException(e);
        }
    }

    public boolean isStepFulfilled() {
        return commandSent;
    }

}

File: devicesetup/src/main/java/io/particle/android/sdk/devicesetup/setupsteps/ConfigureAPStep.kt
package io.particle.android.sdk.devicesetup.setupsteps


import io.particle.android.sdk.devicesetup.commands.CommandClient
import io.particle.android.sdk.devicesetup.commands.ConfigureApCommand
import io.particle.android.sdk.devicesetup.commands.ScanApCommand
import io.particle.android.sdk.devicesetup.commands.data.WifiSecurity
import io.particle.android.sdk.utils.Crypto
import java.io.IOException
import java.security.PublicKey


class ConfigureAPStep internal constructor(stepConfig: StepConfig, private val commandClient: CommandClient,
                                           private val workerThreadApConnector: SetupStepApReconnector,
                                           private val networkToConnectTo: ScanApCommand.Scan,
                                           private val networkSecretPlaintext: String,
                                           private val publicKey: PublicKey) : SetupStep(stepConfig) {
    override val isStepFulfilled: Boolean
        get() = commandSent

    @Volatile
    private var commandSent = false

    @Throws(SetupStepException::class)
    override fun onRunStep() {
        val wifiSecurity = WifiSecurity.fromInteger(networkToConnectTo.wifiSecurityType)
        val builder = ConfigureApCommand.newBuilder()
                .setSsid(networkToConnectTo.ssid)
                .setSecurityType(wifiSecurity)
                .setChannel(networkToConnectTo.channel!!)
                .setIdx(0)
        if (wifiSecurity !== WifiSecurity.OPEN) {
            try {
                builder.setEncryptedPasswordHex(
                        Crypto.encryptAndEncodeToHex(networkSecretPlaintext, publicKey))
            } catch (e: Crypto.CryptoException) {
                // FIXME: try to throw a more specific exception here.
                // Don't throw SetupException here -- if this is failing, it's not
                // going to get any better by the running this SetupStep again, and
                // it can really only fail if the surrounding app code is doing something
                // wrong.  To wit: you *want* the app to crash here (or at least
                // throw out a dialog saying "horrible thing happened!  horrible error
                // code: ..." and then return to a safe "default" activity.
                throw RuntimeException("Error encrypting network credentials", e)
            }

        }
        val command = builder.build()

        try {
            log.d("Ensuring connection to AP")
            workerThreadApConnector.ensureConnectionToSoftAp()

            val response = commandClient.sendCommand(
                    command, ConfigureApCommand.Response::class.java)
            if (!response!!.isOk) {
                throw SetupStepException("Error response code " + response.responseCode +
                        " while configuring device")
            }
            log.d("Configure AP command returned: " + response.responseCode!!)
            commandSent = true

        } catch (e: IOException) {
            throw SetupStepException(e)
        }

    }

}
